src/Corax/Queries/TermMatch.cs

```diff
@@ -273,77 +277,91 @@ static int AndWithVectorizedFunc(ref TermMatch term, Span<long> matches)
                             dstPtr++;
                         }
-
+                    
                         smallerPtr++;
                         if (smallerPtr >= smallerEndPtr)
                             break;
                     }
 
-                    // The scalar version. This shouldnt cost much either way. 
-                    while (smallerPtr < smallerEndPtr && largerPtr < largerEndPtr)
+                    if (largerPtr < largerEndPtr && smallerPtr < smallerEndPtr)
                     {
-                        ulong leftValue = (ulong)*smallerPtr;
-                        ulong rightValue = (ulong)*largerPtr;
-
-                        if (leftValue > rightValue)
-                        {
-                            largerPtr++;
-                        }
-                        else if (leftValue < rightValue)
-                        {
-                            smallerPtr++;
-                        }
-                        else
+                        // The scalar version. This shouldn't cost much either way. 
+                        while (true)
                         {
-                            *dstPtr = (long)leftValue;
-                            dstPtr++;
-                            smallerPtr++;
-                            largerPtr++;
+                            ulong leftValue = (ulong)*smallerPtr;
+                            ulong rightValue = (ulong)*largerPtr;
+
+                            if (leftValue > rightValue)
+                            {
+                                largerPtr++;
+                                if (largerPtr >= largerEndPtr)
+                                    break;
+                            }
+                            else if (leftValue < rightValue)
+                            {
+                                smallerPtr++;
+                                if (smallerPtr >= smallerEndPtr)
+                                    break;
+                            }
+                            else
+                            {
+                                *dstPtr = (long)leftValue;
+                                dstPtr++;
+                                smallerPtr++;
+                                largerPtr++;
+
+                                if (largerPtr >= largerEndPtr || smallerPtr >= smallerEndPtr)
+                                    break;
+                            }
                         }
                     }
 
                     inputPtr = read < matches.Length ? largerPtr : smallerPtr;
 
                     Debug.Assert(inputPtr >= dstPtr);
+                    Debug.Assert((read >= matches.Length ? largerPtr : smallerPtr) - blockStartPtr <= BlockSize);
                 }
 
-                return (int)((ulong)dstPtr - (ulong)Unsafe.AsPointer(ref MemoryMarshal.GetReference(matches))) / sizeof(ulong);
+                return (int)((ulong)dstPtr - (ulong)inputStartPtr) / sizeof(ulong);
```

Comment:
Why are you dividing by `sizeof(ulong)`?  `inputStartPtr` and `dstPtr` are `long*`, not `byte*`.

Reply:
This is a trick to improve codegen (tighter). The JIT is quite bad at generating differences of pointers. What this will do is:

- Take the byte addressable location in memory for `dstPtr` and `inputStartPtr` as a `ulong`
- Calculate the amount of bytes in difference between the 2 locations
- Divide by the size of the element (`ulong` in this case). 
- Cast back to int (mask out the higher part of the register).